File: ts/src/vexfn.ts
import * as readline from 'readline';
import * as http from 'http';
import { VexEvent, VexResponse, VexFnHandler, VexContext, PrimitivesClient } from './types';

function makePrimitivesClient(baseUrl: string | undefined): PrimitivesClient {
  return {
    async call(fnName: string, input: unknown): Promise<unknown> {
      if (!baseUrl) throw new Error('VEX_PRIMITIVES_URL not set');
      const body = JSON.stringify(input);
      return new Promise((resolve, reject) => {
        const url = new URL(`/fn/${fnName}`, baseUrl);
        const req = http.request(
          {
            hostname: url.hostname, port: url.port, path: url.pathname, method: 'POST',
            headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) }
          },
          (res) => {
            let data = '';
            res.on('data', (chunk) => { data += chunk; });
            res.on('end', () => {
              if ((res.statusCode ?? 0) >= 300) {
                reject(new Error(`primitives call ${fnName}: HTTP ${res.statusCode}`));
              } else {
                resolve(JSON.parse(data));
              }
            });
          }
        );
        req.on('error', reject);
        req.write(body);
        req.end();
      });
    }
  };
}

/**
 * Start the stdin/stdout processing loop.
 * Call this from your container's entry point.
 *
 * @example
 * import { run } from '@vextura/vex-sdk';
 * run(async (event, ctx) => ({
 *   eventId: event.eventId,
 *   success: true,
 *   outputData: { approved: true },
 * }));
 */
export async function run(handler: VexFnHandler): Promise<void> {
  const primitivesUrl = process.env['VEX_PRIMITIVES_URL'];
  const primitives = makePrimitivesClient(primitivesUrl);

  const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });

  for await (const line of rl) {
    if (!line.trim()) continue;

    let event: VexEvent;
    try {
      event = JSON.parse(line) as VexEvent;
    } catch (e) {
      const errResp: VexResponse = {
        eventId: '', success: false, errorCode: 'PARSE_ERROR',
        errorMessage: (e as Error).message
      };
      process.stdout.write(JSON.stringify(errResp) + '\n');
      continue;
    }

    const ctx: VexContext = {
      tenantId: event.metadata?.tenantId,
      traceId: event.metadata?.traceId,
      primitives,
    };

    const start = Date.now();
    let resp: VexResponse;
    try {
      resp = await handler(event, ctx);
      resp = { ...resp, eventId: event.eventId, durationMs: Date.now() - start };
    } catch (e) {
      resp = {
        eventId: event.eventId, success: false,
        errorCode: 'HANDLER_ERROR', errorMessage: (e as Error).message
      };
    }

    process.stdout.write(JSON.stringify(resp) + '\n');
  }
}

File: java/src/main/java/io/vextura/sdk/vexfn/VexFn.java
package io.vextura.sdk.vexfn;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.time.Instant;

/**
 * Entry point for a VexEdge function container.
 *
 * <pre>{@code
 * public class Main {
 *     public static void main(String[] args) {
 *         VexFn.run((event, ctx) -> {
 *             String txId = (String) event.inputData().get("transactionId");
 *             return VexResponse.ok(event.eventId(), Map.of("approved", true), 0L);
 *         });
 *     }
 * }
 * }</pre>
 */
public final class VexFn {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private VexFn() {}

    /**
     * Start the stdin/stdout processing loop.
     * Blocks until stdin is closed (container shutdown).
     */
    public static void run(VexFnHandler handler) {
        String primUrl  = System.getenv("VEX_PRIMITIVES_URL");
        PrimitivesClient primitives = primUrl != null ? new PrimitivesClient(primUrl) : null;

        PrintStream out = System.out;
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

        String line;
        try {
            while ((line = in.readLine()) != null) {
                if (line.isBlank()) continue;

                VexEvent event;
                try {
                    event = MAPPER.readValue(line, VexEvent.class);
                } catch (Exception e) {
                    writeResponse(out, VexResponse.error("", "PARSE_ERROR", e.getMessage()));
                    continue;
                }

                Context ctx = new Context(
                        primitives,
                        event.metadata() != null ? event.metadata().tenantId() : null,
                        event.metadata() != null ? event.metadata().traceId()  : null
                );

                long start = Instant.now().toEpochMilli();
                VexResponse resp;
                try {
                    resp = handler.handle(event, ctx);
                    if (resp == null) {
                        resp = VexResponse.ok(event.eventId(), null, 0L);
                    }
                    long dur = Instant.now().toEpochMilli() - start;
                    resp = new VexResponse(
                            event.eventId(), resp.success(), resp.outputData(),
                            resp.errorCode(), resp.errorMessage(), dur);
                } catch (Exception e) {
                    resp = VexResponse.error(event.eventId(), "HANDLER_ERROR", e.getMessage());
                }

                writeResponse(out, resp);
            }
        } catch (Exception e) {
            System.err.println("vexfn: fatal: " + e.getMessage());
        }
    }

    private static void writeResponse(PrintStream out, VexResponse resp) {
        try {
            out.println(MAPPER.writeValueAsString(resp));
            out.flush();
        } catch (Exception e) {
            System.err.println("vexfn: encode error: " + e.getMessage());
        }
    }
}
